feat: add missing schema validation

quesebifurcan opened the pull request:
Validate extracted entities against provided schemas. 

In `aether.kernel.api.views.validate_mappings`,
differentiate between an invalid request (400) and an
unforeseen extraction exception (500).

Add a MappingValidationSerializer.

Add tests for entity validation.

Example submission response for validation errors:
```
{
    "id": "6c6579bb-0367-4b79-9944-63de62b3aab4",
    "url": "http://kernel.aether.local:8000/submissions/6c6579bb-0367-4b79-9944-63de62b3aab4/",
    "mapping_url": "http://kernel.aether.local:8000/mappings/ed0883ea-23f1-4afc-b0a1-8a7e8d21d70d/",
    "entities_url": "http://kernel.aether.local:8000/entities/?submission=6c6579bb-0367-4b79-9944-63de62b3aab4",
    "attachments_url": "http://kernel.aether.local:8000/attachments/?submission=6c6579bb-0367-4b79-9944-63de62b3aab4",
    "attachments": [],
    "created": "2018-05-16T08:38:22.244513Z",
    "modified": "2018-05-16T08:38:22.272596Z",
    "revision": "1",
    "map_revision": "1",
    "date": "2018-05-16T08:38:22.272630Z",
    "payload": {
        "firstname": 1,
        "lastname": "b",
        "aether_errors": [
            {
                "type": "EntityValidationError",
                "description": "Extracted record did not conform to registered schema",
                "data": {
                    "id": "c28e0ed9-c628-4b60-91f9-b3337a807dd8",
                    "firstName": 1,
                    "familyName": "b"
                }
            }
        ],
        "aether_extractor_enrichment": {
            "Person": {
                "id": [
                    "c28e0ed9-c628-4b60-91f9-b3337a807dd8"
                ]
            }
        }
    },
    "mapping": "ed0883ea-23f1-4afc-b0a1-8a7e8d21d70d"
}
```

Resolves: https://jira.ehealthafrica.org/browse/AET-242

quesebifurcan committed:
add ipython and ipdb

quesebifurcan committed:
add tests for validate_entity()

quesebifurcan committed:
WIP validation

quesebifurcan committed:
doc: comments

obdulia-losantos commented:
LGTM!

obdulia-losantos commented:
Could it be possible to include a better error handling when the mapping does not include an `id` for the used entities? So far is an obscure 400 error.

Not sure if here or in another PR...

quesebifurcan commented:
@obdulia-losantos Yes, good point, we should fix that. I'd prefer to address that in a separate PR, if that's ok with you?

tooxie commented on aether-kernel/conf/pip/primary-requirements.txt:
```diff
@@ -22,6 +22,8 @@ drf-dynamic-fields
 jsonpath_ng
 jsonschema
+ipdb
+ipython
```
Are these dependencies going to be deployed to production as well?

quesebifurcan replied:
Good point and yes, at the moment they will be. Perhaps we should keep them out of this file until we have a separate requirements file for development?

tooxie replied:
Yes, I would prefer that :)

tooxie commented on aether-kernel/aether/kernel/api/views.py:
```diff
@@ -174,27 +174,24 @@ def validate_mappings(request):
+        except Exception as e:
+            return Response(str(e), status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+    else:
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```
You could return early if the serializer is not valid and save 1 indentation level. Remember: *flat is better than nested* 😋 

Also, it seems to me that there are too many lines in that `try`. Is there any logging done of the errors besides sending them to the client?

quesebifurcan committed:
remove ipython and ipdb

quesebifurcan committed:
shorten try statement

quesebifurcan committed:
add comment about return values

tooxie commented on aether-kernel/aether/kernel/api/tests/test_mapping_validation.py:
```diff
@@ -158,16 +158,20 @@ def test_validate_mapping__failure(self):
+                ),
+            ).as_dict(),
+            mapping_validation.JsonpathValidationError.from_failure(
+                mapping_validation.Failure(
```
Why are `path` and `description` wrapped in a `Failure` object just to be then unwrapped by `JsonpathValidationError.from_failure()`? You could directly instantiate it:
```py
JsonpathValidationError(
    path=mapping_definition['mapping'][1][0],
    description=mapping_validation.MESSAGE_NO_MATCH
)
```

tooxie commented on aether-kernel/aether/kernel/api/mapping_validation.py:
```diff
@@ -23,10 +23,37 @@
+Failure = collections.namedtuple('Failure', ['path', 'description'])
+
+
+class JsonpathValidationError(object):
```
I would have expected it to extend `Exception` given the name. And looking at how it's used, it looks like a function could have been enough:
```py
def failure_to_dict(failure):
    return {
        'description': failure.description,
        'path': failure.path,
        'type': 'JsonpathValidationError',
    }
```

Have you considered that approach?

tooxie commented on aether-kernel/aether/kernel/api/utils.py:
```diff
@@ -401,7 +416,32 @@ def extract_entity(entity_type, entities, requirements, data, entity_stub):
+                validated_entities[entity_name].append(entity_payload)
+            except EntityValidationError as err:
+                error = {
+                    'type': EntityValidationError.__name__,
```
Why not just use a string literal here? The name will always be *'EntityValidationError'*, right?

tooxie commented on aether-kernel/aether/kernel/api/views.py:
```diff
@@ -174,27 +198,18 @@ def validate_mappings(request):
-            status=status.HTTP_400_BAD_REQUEST
-        )
+        error = str(e)
+        logger.exception(error)
```
It would be very useful to log the traceback as well as the error message.

quesebifurcan committed:
WIP new validation

quesebifurcan committed:
include traceback in log

quesebifurcan committed:
Merge branch 'develop' into feat/add-missing-schema-validation

quesebifurcan commented:
Summary of updates:

1. Removed confusing `JsonpathValidationError` class. As Alvaro pointed out, it looked too much like an exception.
2. Include traceback in error log.
3. Change severity level of log message from `exception` to `debug`.
4. Rewrite `validate_setter()` to handle flat destination schemas (following up on today's discussion in #aether-dev).
5. (Finally) allow multiple validation errors (entity validation errors, jsonpath lookup errors) to be returned from the `validate-mappings` endpoint.
6. Removed confusing "type" information in responses (the descriptions provide enough information). It would be nice to have but is not a requirement for now. We can revisit this as part of https://jira.ehealthafrica.org/browse/AET-262.

@tooxie

tooxie commented on aether-kernel/aether/kernel/api/tests/test_views.py:
```diff
@@ -420,10 +429,22 @@ def test_example_entity_extraction__failure(self):
+        expected = set([
+                'Could not find schema "person"',
+                'No match for path',
+                'Extracted record did not conform to registered schema',
```
A couple of unimportant nitpicks:
* Why do you turn the lists into sets? You could just compare the lists.
* Extra spacing :)

quesebifurcan replied:
Ah, good catch. The `set()` thing is just habit, really. I'll fix that along with the spacing (why didn't the linter catch that? :confused:)

tooxie reviewed (dismissed):
Feel free to ignore my comments, I think the PR is good to go as is.

quesebifurcan committed:
chore: lint

quesebifurcan committed:
remove logging statement
